02-peer-protocol.md

```diff
@@ -1713,6 +1762,161 @@ satoshis, which is possible if `dust_limit_satoshis` is below 546 satoshis).
+  - MUST set `fee_satoshis` so that at least one output is not dust.
+  - MUST use the last sent and received `shutdown.scriptpubkey` to generate the closing transaction specified in [BOLT #3](03-transactions.md#closing-transaction).
+  - MUST set `locktime` to the desired `nLockTime` of the closing transaction.
+  - MUST set `signature` fields as valid signature using its `funding_pubkey` of:
+    - `closer_output_only`: closing transaction with only the local ("closer") output.
+    - `closee_output_only`: closing transaction with only the remote ("closee") output.
+    - `closer_and_closee_outputs`: closing transaction with both the closer and closee outputs.
+  - If the local outstanding balance (in millisatoshi) is less than the remote outstanding balance:
+    - MUST NOT set `closer_output_only`.
+    - MUST set `closee_output_only` if the local output amount is dust.
+    - MAY set `closee_output_only` if it considers the local output amount uneconomical AND its `scriptpubkey` is not `OP_RETURN`.
+  - Otherwise (not lesser amount, cannot remove own output):
+    - MUST NOT set `closee_output_only`.
+    - If the closee's output amount is dust:
+      - MUST set `closer_output_only`.
+      - SHOULD NOT set `closer_and_closee_outputs`.
+    - Otherwise:
+      - MUST set both `closer_output_only` and `closer_and_closee_outputs`.
+  - If it wants to send another `closing_complete` (e.g. with a different `fee_satoshis`):
+    - MUST send `shutdown` first.
+    - MUST receive `shutdown` from the remote node in response.
+    - MUST use the `scriptpubkey`s from those `shutdown` messages.
+
+The receiver of `closing_complete` (aka. "the closee"):
+  - If `fee_satoshis` is greater than the closer's outstanding balance:
+    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
+  - Select a signature for validation:
+    - if the local output amount is dust:
+      - MUST use `closer_output_only`.
+    - otherwise, if it considers the local output amount uneconomical AND its `scriptpubkey` is not `OP_RETURN`:
+      - MUST use `closer_output_only`.
+    - otherwise, if `closer_and_closee_outputs` is present:
+      - MUST use `closer_and_closee_outputs`.
+    - otherwise:
+      - MUST use `closee_output_only`.
+  - If the selected signature field does not exist:
+    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
+  - If the signature field is not valid for the corresponding closing transaction specified in [BOLT #3](03-transactions.md#closing-transaction):
+    - MUST ignore `closing_complete`.
+  - If the signature field is non-compliant with LOW-S-standard rule<sup>[LOWS](https://github.com/bitcoin/bitcoin/pull/6769)</sup>:
+    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
+  - MUST sign and broadcast the corresponding closing transaction.
```

Comment:
I referring to the race condition here,
```
 |       |             closing_complete |       |
 |       |          <-------------------|       |
 |       |                              |       |
 |       |  <Alice updates her script>  |       | (*) This is a concurrent update while Bob is sending closing_complete
 |       |                              |       |
 |       | shutdown(scriptA3)           |       |
 |       |------------------->          |       |
 |       |   closing_complete           |       |
 |       |<-------------------          |       | (*) A doesn't answer with closing_sig because B's sig doesn't use scriptA3
```

But now I see Alice MUST ignore the `closing_complete` due to sig being invalid so nvm.